File: artifacts/quadtree-0.1.0/decompose/src/bounds.c
void quadtree_bounds_extend(quadtree_bounds_t *bounds, double x, double y)
{
  bounds->nw->x = fmin(x, bounds->nw->x);
  bounds->nw->y = fmax(y, bounds->nw->y);
  bounds->se->x = fmax(x, bounds->se->x);
  bounds->se->y = fmin(y, bounds->se->y);
  bounds->width = fabs(bounds->nw->x - bounds->se->x);
  bounds->height = fabs(bounds->nw->y - bounds->se->y);
}


/*** DEPENDENCIES:
typedef struct quadtree_bounds
{
  quadtree_point_t *nw;
  quadtree_point_t *se;
  double width;
  double height;
} quadtree_bounds_t
----------------------------
***/


void quadtree_bounds_free(quadtree_bounds_t *bounds)
{
  quadtree_point_free(bounds->nw);
  quadtree_point_free(bounds->se);
  free(bounds);
}


/*** DEPENDENCIES:
typedef struct quadtree_bounds
{
  quadtree_point_t *nw;
  quadtree_point_t *se;
  double width;
  double height;
} quadtree_bounds_t
----------------------------
void quadtree_point_free(quadtree_point_t *point)
{
  free(point);
}


----------------------------
***/


quadtree_bounds_t *quadtree_bounds_new()
{
  quadtree_bounds_t *bounds;
  unsigned int bounds_idx = 0;
  if ((bounds_idx = malloc(sizeof(*bounds_idx))) == 0)
  {
    return 0;
  }
  bounds->nw = quadtree_point_new(INFINITY, -INFINITY);
  bounds->se = quadtree_point_new(-INFINITY, INFINITY);
  bounds->width = 0;
  bounds->height = 0;
  return bounds;
}


/*** DEPENDENCIES:
typedef struct quadtree_bounds
{
  quadtree_point_t *nw;
  quadtree_point_t *se;
  double width;
  double height;
} quadtree_bounds_t
----------------------------
quadtree_point_t *quadtree_point_new(double x, double y)
{
  quadtree_point_t *point;
  unsigned int point_idx = 0;
  if (!(point_idx = malloc(sizeof(*point_idx))))
  {
    return 0;
  }
  point->x = x;
  point->y = y;
  return point;
}


----------------------------
***/

File: artifacts/quadtree-0.1.0/translate/src/bounds.rs
use crate::*;
use std::f64;
use std::f64::INFINITY;
use std::f64::NEG_INFINITY;


pub fn quadtree_bounds_extend(bounds: &mut QuadtreeBounds, x: f64, y: f64) {
    // Handle nw updates
    if let Some(ref mut nw) = bounds.nw {
        nw.x = f64::min(x, nw.x);
        nw.y = f64::max(y, nw.y);
    }

    // Handle se updates
    if let Some(ref mut se) = bounds.se {
        se.x = f64::max(x, se.x);
        se.y = f64::min(y, se.y);
    }

    // Calculate new dimensions
    if let (Some(nw), Some(se)) = (&bounds.nw, &bounds.se) {
        bounds.width = f64::abs(nw.x - se.x);
        bounds.height = f64::abs(nw.y - se.y);
    }
}

pub fn quadtree_bounds_new() -> Option<Box<QuadtreeBounds>> {
    let nw = quadtree_point_new(INFINITY, NEG_INFINITY);
    let se = quadtree_point_new(NEG_INFINITY, INFINITY);

    Some(Box::new(QuadtreeBounds {
        nw,
        se,
        width: 0.0,
        height: 0.0,
    }))
}
pub fn quadtree_bounds_free(bounds: Option<Box<QuadtreeBounds>>) {
    if let Some(bounds) = bounds {
        quadtree_point_free(bounds.nw);
        quadtree_point_free(bounds.se);
        // The Box will be automatically dropped here, freeing the memory
    }
}
